File: src/main/java/fr/marstech/mtlinkspray/controller/ThymeleafViewControllerInterface.java
package fr.marstech.mtlinkspray.controller;

import fr.marstech.mtlinkspray.enums.ViewNameEnum;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.servlet.ModelAndView;

import static fr.marstech.mtlinkspray.utils.NetworkUtils.getHeadersAsMap;
import static java.text.MessageFormat.format;

public interface ThymeleafViewControllerInterface {

    ModelAndView getModelAndView();

    default ModelAndView getModelAndView(ViewNameEnum viewNameEnum) {
        return new ModelAndView(viewNameEnum.getViewName()).addObject("viewNameEnum", viewNameEnum);
    }

    default ModelAndView getModelAndView(ViewNameEnum viewNameEnum, HttpServletRequest httpServletRequest) {
        return new ModelAndView(viewNameEnum.getViewName()).addObject("viewNameEnum", viewNameEnum).addObject("headers", getHeadersAsMap(httpServletRequest));
    }

    default ModelAndView getModelAndViewToForward(String forwardUrl) {
        return new ModelAndView(format("forward:{0}", forwardUrl));
    }

    default String getRedirectUrl(String url) {
        return format("redirect:{0}", url);
    }
}

File: src/main/kotlin/fr/marstech/mtlinkspray/controller/view/ThymeleafViewControllerInterface.kt
package fr.marstech.mtlinkspray.controller.view

import fr.marstech.mtlinkspray.enums.ViewNameEnum
import fr.marstech.mtlinkspray.utils.NetworkUtils.getHeadersAsMap
import jakarta.servlet.http.HttpServletRequest
import org.springframework.web.servlet.ModelAndView

/**
 * Interface for controllers returning Thymeleaf views.
 */
interface ThymeleafViewControllerInterface {

    fun getModelAndView(): ModelAndView?

    /**
     * Returns a ModelAndView for the given view name.
     */
    fun getModelAndView(viewNameEnum: ViewNameEnum): ModelAndView =
        ModelAndView(viewNameEnum.viewName).addObject("viewNameEnum", viewNameEnum)

    /**
     * Returns a ModelAndView for the given view name and request headers.
     */
    fun getModelAndView(viewNameEnum: ViewNameEnum, httpServletRequest: HttpServletRequest): ModelAndView =
        ModelAndView(viewNameEnum.viewName)
            .addObject("viewNameEnum", viewNameEnum)
            .addObject("headers", getHeadersAsMap(httpServletRequest))

    /**
     * Returns a ModelAndView to forward to the given URL.
     */
    fun getModelAndViewToForward(forwardUrl: String): ModelAndView =
        ModelAndView("forward:$forwardUrl")

    /**
     * Returns a ModelAndView to redirect to the given URL.
     */
    fun getModelAndViewToRedirect(redirectUrl: String): ModelAndView =
        ModelAndView(getRedirectUrl(redirectUrl))

    /**
     * Returns a redirect URL string.
     */
    fun getRedirectUrl(url: String): String = "redirect:$url"

}